subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1087,6 +1093,29 @@ public void preventFromFurtherMutation() {
             configurationAttributes = new ImmutableAttributeContainerWithErrorMessage(delegatee, this.displayName);
             outgoing.preventFromFurtherMutation();
             canBeMutated = false;
+
+            if (isCanBeConsumed()) {
+                ensureUniqueAttributesOnConsumableConfigurations();
+            }
+        }
+    }
+
+    private void ensureUniqueAttributesOnConsumableConfigurations() {
+        if (configurationsProvider != null) {
+            final Set<? extends ConfigurationInternal> all = configurationsProvider.getAll();
+            if (all != null) {
+                final Set<? extends ConfigurationInternal> potentialDuplicates = configurationsProvider.getAll().stream()
+                    .filter(c -> c != this)
+                    .filter(c -> !c.isMutable())
+                    .filter(c -> c.isCanBeConsumed())
+                    .collect(Collectors.toSet());
+
+                for (ConfigurationInternal other : potentialDuplicates) {
+                    if (!other.getAttributes().isEmpty() && !getAttributes().isEmpty() && other.getAttributes().asMap().equals(getAttributes().asMap())) {
+                        throw new GradleException("Consumable configurations within a project must have unique attribute sets, but " + getDisplayName() + " and " + other.getDisplayName() + " contain identical attribute sets.");
+                    }
```

Comment:
1. Everything can be skipped if this configuration's attributes are empty.
2. Potential duplicate configurations can pre-filter out any configurations without attributes.
3. Collecting the potential duplicates can be skipped by using `forEach` on a Stream.